wade-review.md

```diff
@@ -0,0 +1,596 @@
+	static final String STORE_KEY = "state-counter";
+   @Override
+   public Processor<String, GitHubPRInfo, String, GitHubPRStateCounter> get() {
+		return new Processor<>() {
+
+    private KeyValueStore<String, GitHubPRStateCounter> kvStore;
+                @Override
+                public void init(final ProcessorContext<String, GitHubPRStateCounter> context) {
+                    this.kvStore = context.getStateStore(STORE_NAME);
+
+                    context.schedule(Duration.ofSeconds(1), PunctuationType.STREAM_TIME, timestamp -> {
+                        GitHubPRStateCounter entry = kvStore.get(STORE_KEY);
+                        System.out.printf("Store value %s%n", entry);
+                        context.forward(new Record<>("pr", entry, timestamp));
+                    });
+	}
+}
+
+```
+
+Then, I’ll initialize it with our state store. The init method schedules a punctuation to fire every second, printing the ratio in the store. 
+
+```java
+static class MyProcessorSupplier implements ProcessorSupplier<String, GitHubPRInfo, String, GitHubPRStateCounter> { 
+	static final String STORE_KEY = "state-counter";
+   @Override
+   public Processor<String, GitHubPRInfo, String, GitHubPRStateCounter> get() {
+		return new Processor<>() {
+
+    private KeyValueStore<String, GitHubPRStateCounter> kvStore;
+                @Override
+                public void init(final ProcessorContext<String, GitHubPRStateCounter> context) {
+                    this.kvStore = context.getStateStore(STORE_NAME);
+
+                    context.schedule(Duration.ofSeconds(1), PunctuationType.STREAM_TIME, timestamp -> {
+                        GitHubPRStateCounter entry = kvStore.get(STORE_KEY);
+                        System.out.printf("Store value %s%n", entry);
+                        context.forward(new Record<>("pr", entry, timestamp));
+                    });
+	}
+              @Override
+                public void process(final Record<String, GitHubPRInfo> record) {
+
+                }
+            };
+        }
+}
+```
+
+Next, I’ll write the process method to  take in the events, mark them as open or closed, increment the count, and stash them in the state store.
+
+```java
+static class MyProcessorSupplier implements ProcessorSupplier<String, GitHubPRInfo, String, GitHubPRStateCounter> { 
+	static final String STORE_KEY = "state-counter";
+   @Override
+   public Processor<String, GitHubPRInfo, String, GitHubPRStateCounter> get() {
+		return new Processor<>() {
+
+    private KeyValueStore<String, GitHubPRStateCounter> kvStore;
+                @Override
+                public void init(final ProcessorContext<String, GitHubPRStateCounter> context) {
+                    this.kvStore = context.getStateStore(STORE_NAME);
+
+                    context.schedule(Duration.ofSeconds(1), PunctuationType.STREAM_TIME, timestamp -> {
+                        GitHubPRStateCounter entry = kvStore.get(STORE_KEY);
```

Comment:
My gut here is to say that the `init` function is where the store is initialized, and that the `process` function is the point of entry for new records from the Connector so that is where we should check for `null` first. But I'd like to ask @bbejeck to take a look at this comment as well.

Reply:
I'm making an assumption here, but typically, "init" would be called at the very beginning of creating a thing. I'm assuming the basic flow of this thing would be something like:

```
init()
process()
process()
...
```
It would seem odd if the flow was:
```
process()
init()
process()
...
```
So I would expect that the first place to check for null is in the `init` function. But given that you aren't checking for null there, and I assume your code probably works, is it possible that you don't even need the null check? If you remove it does anything break?

Maybe the `getStateStore` function already does all of that work for you and this is redundant.

Reply:
The `null` check in the process function is due to the fact that until a record comes in, there will be no `GitHubPRStateCounter` entry in the store. There's no check in the punctuation code because it's assuming the topology has already processed some records.  Since the punctuation is using stream time, you can be sure that the topology has processed something, hence the counter is in the store.  But adding an extra `null` check wouldn't hurt. The `init` method is called by Kafka Streams on startup.

Reply:
Ah...My bad. I missed the fact that this code is inside of a scheduled lambda function, rather than inside the init itself.

That makes a lot more sense. Would be interesting to discuss those details in the video.